### PR TITLE
Make sure we resolve the symlink of the cwd

### LINKED ISSUE #3015: Crash when the current directory is a symlink to a directory in a GIT repo
**Describe the bug**
Crash when the current directory is a symlink to a directory in a GIT repo
**To Reproduce**
```
% cd /tmp
% mkdir repotest
% cd repotest
% git init
Initialized empty Git repository in /tmp/repotest/.git/
% mkdir foobar
% git init
% touch foobar/blah
% git add .
% git commit -m "test ignore" -a
[main (root-commit) c8bccd0] test ignore
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 foobar/blah
% ln -s /tmp/repotest/foobar /tmp
% cd /tmp/foobar
% lazygit
2023/09/20 09:55:27 An error occurred! Please create an issue at: https://github.com/jesseduffield/lazygit/issues

*errors.errorString Must open lazygit in a git repository
github.com/jesseduffield/lazygit/pkg/commands/git.go:235 (0x55818ecc1da6)
github.com/jesseduffield/lazygit/pkg/commands/git.go:87 (0x55818ecc0936)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:271 (0x55818edf06eb)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:656 (0x55818edf32db)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:675 (0x55818edf39c5)
github.com/jesseduffield/lazygit/pkg/utils/utils.go:108 (0x55818eb4f0fc)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:674 (0x55818edf390e)
github.com/jesseduffield/lazygit/pkg/app/app.go:263 (0x55818ee1f57d)
github.com/jesseduffield/lazygit/pkg/app/app.go:48 (0x55818ee1f512)
github.com/jesseduffield/lazygit/pkg/app/entry_point.go:150 (0x55818ee216ea)
github.com/jesseduffield/lazygit/main.go:23 (0x55818ee22e78)
runtime/internal/atomic/types.go:194 (0x55818e77ceb2)
runtime/asm_amd64.s:1650 (0x55818e7ad121)
```

**Expected behavior**
no crash or at least a error message maybe :)

**Screenshots**

**Version info:**
commit=v0.40.2, build date=2023-08-12T17:47:33Z, build source=binaryRelease, version=0.40.2, os=linux, arch=amd64, git version=2.42.0

**Additional context**

arch package fwiw

**Note:** please try updating to the latest version or [manually building](https://github.com/jesseduffield/lazygit/#manual) the latest `master` to see if the issue still occurs.

### PR DESCRIPTION
**Pull request by chmouel**
Make sure we resolve the current directory symlink or it would fail
getting confused about the repo root.

Fixes #3015 

* [X] Cheatsheets are up-to-date (run `go run scripts/cheatsheet/main.go generate`)
* [X] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [ ] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)

### CODE REVIEW

**Comment by jesseduffield** [comment]
@chmouel yep it is annoying that aeofs doesn't cover symlinking. I say in that case it's best to create an integration test for this. Should be simple enough to verify that you can start lazygit inside the symlinked repo without it exiting with an error. There's an example of such a test at pkg/integration/tests/config/remote_named_star.go

**Comment by mark2185** [comment]
This works only if `lazygit` is launched from a folder that is a symlink, it doesn't work if it's e.g. one level deeper.

E.g.:
```
$> file ~/.config/foo
/home/mark/.config/foo: symbolic link to /home/mark/workspace/gits/dotfiles/.config/foo
$> cd ~/.config/foo
$> lazygit
(works)
$> cd foo-deeper/
$> lazygit
2023/10/18 16:06:57 An error occurred! Please create an issue at: https://github.com/jesseduffield/lazygit/issues

*errors.errorString Must open lazygit in a git repository
/home/mark/workspace/gits/lazygit/pkg/commands/git.go:236 (0x985906)
	findWorktreeRoot: return "", errors.New("Must open lazygit in a git repository")
/home/mark/workspace/gits/lazygit/pkg/commands/git.go:90 (0x984488)
	NewGitCommand: rootDirectory, err := findWorktreeRoot(cmn.Fs, currentPath)
/home/mark/workspace/gits/lazygit/pkg/gui/gui.go:271 (0xabf68b)
	(*Gui).onNewRepo: gui.git, err = commands.NewGitCommand(
/home/mark/workspace/gits/lazygit/pkg/gui/gui.go:655 (0xac223b)
	(*Gui).Run: if err := gui.onNewRepo(startArgs, context.NO_CONTEXT); err != nil {
/home/mark/workspace/gits/lazygit/pkg/gui/gui.go:674 (0xac2925)
	(*Gui).RunAndHandleError.func1: if err := gui.Run(startArgs); err != nil {
/home/mark/workspace/gits/lazygit/pkg/utils/utils.go:111 (0x810bdc)
	SafeWithError: err := f()
/home/mark/workspace/gits/lazygit/pkg/gui/gui.go:673 (0xac286e)
	(*Gui).RunAndHandleError: return utils.SafeWithError(func() error {
/home/mark/workspace/gits/lazygit/pkg/app/app.go:265 (0xaeea1d)
	Run: err := app.Gui.RunAndHandleError(startArgs)
/home/mark/workspace/gits/lazygit/pkg/app/app.go:48 (0xaee9b2)
	Run: err = app.Run(startArgs)
/home/mark/workspace/gits/lazygit/pkg/app/entry_point.go:150 (0xaf0baa)
	Start: Run(appConfig, common, appTypes.NewStartArgs(cliArgs.FilterPath, parsedGitArg, integrationTest))
/home/mark/workspace/gits/lazygit/main.go:23 (0xaf2338)
	main: app.Start(ldFlagsBuildInfo, nil)
/usr/lib/go/src/runtime/internal/atomic/types.go:194 (0x43d8bb)
	(*Uint32).Load: return Load(&u.value)
/usr/lib/go/src/runtime/asm_amd64.s:1650 (0x46d961)
	goexit: BYTE	$0x90	// NOP

**Comment by JL102** [comment]
Before discovering this PR, I created a minimal repo to reproduce the same symlink issue. Feel free to use it if it helps. https://github.com/JL102/lazygit-submodule-symlink-test/tree/main

File: pkg/commands/git_commands/repo_paths.go [hunk 1]
```diff
@@ -228,7 +228,7 @@ func getCurrentRepoGitDirPath(
 }
 
 // takes a path containing a symlink and returns the true path
-func resolveSymlink(path string) (string, error) {
+func ResolveSymlink(path string) (string, error) {
 	l, err := os.Lstat(path)
```
Comment (mark2185) [inline pkg/commands/git_commands/repo_paths.go:232]: Why do we even need this check if it's `os.ModeSymlink`?

In order to be able to call `lazygit` from a folder whose *path* _contains_ a symlink this check fails, and `lazygit` fails to start, but if we just `return filepath.EvalSymlinks(path)`, all works as expected.

The example is what I've written in the conversation for this PR.

@jesseduffield could this have been done for potential performance reasons? Tried blaming it, but it went through two refactors so I stopped.

**Comment by hboetes** [comment]
Works for me.  👍

File: pkg/commands/git_commands/repo_paths.go (hunk 1 above)
Comment (jesseduffield) [reply pkg/commands/git_commands/repo_paths.go:232]: I've never made any performance-related decisions around symlinks. If `filepath.EvalSymlinks(path)` works I'm happy for us to use it

Comment (mark2185) [reply pkg/commands/git_commands/repo_paths.go:232]: The docs say:
```
// EvalSymlinks returns the path name after the evaluation of any symbolic
// links.
// If path is relative the result will be relative to the current directory,
// unless one of the components is an absolute symbolic link.
// EvalSymlinks calls Clean on the result.
```

That would suggest nothing changes if the given path _isn't_ a symlink, so I see no reason to `if-else` on it being a `symlink`.

But I'd wait for tests that confirm it.

**Comment by mark2185** [comment]
@chmouel are you still up for this? It would be great to have an integration test for this, and like Jesse said:

> I say in that case it's best to create an integration test for this. Should be simple enough to verify that you can start lazygit inside > the symlinked repo without it exiting with an error. There's an example of such a test at pkg/integration/tests/config/remote_named_star.go

**Comment by chmouel** [comment]
@mark2185 you are right, i am still up for it, let me find some time this week-end or i'll close this PR

**Comment by chmouel** [comment]
So I have tried to build an integration and I have some difficulty changing the directory of where the test run:

```go
//cat pkg/integration/tests/misc/symlinks_repo.go
package misc
 
import (
    "fmt"
    "os"
    "path/filepath"
 
    "github.com/jesseduffield/lazygit/pkg/config"
    . "github.com/jesseduffield/lazygit/pkg/integration/components"
)
 
var SymlinkRepo = NewIntegrationTest(NewIntegrationTestArgs{
    Description:  "Make sure lazygit open on a symlink into a repository",
    ExtraCmdArgs: []string{},
    Skip:         false,
    SetupConfig:  func(config *config.AppConfig) {},
    SetupRepo: func(shell *Shell) {
        shell.NewBranch("mybranch")
        shell.CreateDir("symlinked")
        shell.CreateFileAndAdd("symlinked/blah.txt", "blah")
        shell.Commit("initial commit")
    },
    Run: func(t *TestDriver, _ config.KeybindingConfig) {
        tempdir, _ := os.MkdirTemp("", "lazygit-symlink")
        t.Shell().RunShellCommand(fmt.Sprintf("ln -s $PWD/symlinked %s", tempdir))
        _ := os.Chdir(filepath.Join(tempdir, "symlinked"))
         t.Views().Commits()
    },
})
```

It need to be outside of where the test run currently since that bug would not reproduce otherwise so using a tempdir, but when I run that test against `master` it doesn't fail, i can't figure out why the Chdir is not taken into account, if someone has an idea that would be great

**Comment by jwhitley** [comment]
FYI, this PR would be obsoleted by PR #3183, which replaces lazygit's pathfinding code by delegating that work to git (via `git rev-parse`). I've confirmed this by following @JL102's reproduction steps in the test repo linked above. In short, since git already correctly handles many edge cases of this ilk, delegating to git is ultimately a more robust approach.

**Comment by chmouel** [comment]
@jwhitley perfect thank you, i may close this PR now since it was stalled for too long!
